File: web/tests/tag-utils.test.mjs
import test from 'node:test';
import assert from 'node:assert/strict';
import { canonicalTag, parseTagsInput, tagMatchesQuery } from '../js/tag-utils.js';

test('known spelling variants collapse to one canonical tag', () => {
  assert.equal(canonicalTag('BBQ'), 'bbq');
  assert.equal(canonicalTag('#BBQ'), 'bbq');
  assert.equal(canonicalTag('Barbecue'), 'bbq');
  assert.equal(canonicalTag('barbeque'), 'bbq');
  assert.equal(canonicalTag('Bar-B-Q'), 'bbq');
});

test('distinct concepts are not merged', () => {
  assert.equal(canonicalTag('vegan'), 'vegan');
  assert.equal(canonicalTag('vegetarian'), 'vegetarian');
});

test('parseTagsInput splits, dedupes and canonicalizes', () => {
  assert.deepEqual(parseTagsInput('#BBQ, camping   Camping bbq'), ['bbq', 'camping']);
});

test('parseTagsInput caps at maxTags', () => {
  const raw = Array.from({ length: 12 }, (_, i) => `tag${i}`).join(' ');
  assert.equal(parseTagsInput(raw).length, 8);
});

test('tagMatchesQuery handles aliases and partial input', () => {
  assert.equal(tagMatchesQuery('bbq', 'barbecue'), true);
  assert.equal(tagMatchesQuery('bbq', '#bbq'), true);
  assert.equal(tagMatchesQuery('bbq', 'bb'), true);
  assert.equal(tagMatchesQuery('bbq', 'camping'), false);
});

File: app/src/test/java/com/chefvoice/app/util/TagUtilsTest.kt
package com.chefvoice.app.util

import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Test

class TagUtilsTest {
    @Test
    fun knownSpellingVariantsCollapseToOneCanonicalTag() {
        assertEquals("bbq", canonicalTag("BBQ"))
        assertEquals("bbq", canonicalTag("#BBQ"))
        assertEquals("bbq", canonicalTag("Barbecue"))
        assertEquals("bbq", canonicalTag("barbeque"))
        assertEquals("bbq", canonicalTag("Bar-B-Q"))
    }

    @Test
    fun distinctConceptsAreNotMerged() {
        assertEquals("vegan", canonicalTag("vegan"))
        assertEquals("vegetarian", canonicalTag("vegetarian"))
    }

    @Test
    fun parseTagsInputSplitsDedupesAndCaps() {
        val tags = parseTagsInput("#BBQ, camping   Camping bbq")
        assertEquals(listOf("bbq", "camping"), tags)
    }

    @Test
    fun parseTagsInputCapsAtMaxTags() {
        val raw = (1..12).joinToString(" ") { "tag$it" }
        assertEquals(8, parseTagsInput(raw).size)
    }

    @Test
    fun tagMatchesQueryHandlesAliasesAndPartials() {
        assertTrue(tagMatchesQuery("bbq", "barbecue"))
        assertTrue(tagMatchesQuery("bbq", "#bbq"))
        assertTrue(tagMatchesQuery("bbq", "bb"))
        assertFalse(tagMatchesQuery("bbq", "camping"))
    }
}
